akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/ClusterShardAllocationMixin.scala

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2020-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding.internal
+
+import java.lang.{ Integer => JInteger }
+import java.lang.{ Boolean => JBoolean }
+
+import scala.collection.immutable
+
+import akka.actor.ActorRef
+import akka.actor.Address
+import akka.annotation.InternalApi
+import akka.cluster.ClusterEvent.CurrentClusterState
+import akka.cluster.Member
+import akka.cluster.MemberStatus
+import akka.cluster.sharding.ShardRegion.ShardId
+import akka.cluster.sharding.internal.ClusterShardAllocationMixin.JoiningCluster
+import akka.cluster.sharding.internal.ClusterShardAllocationMixin.RegionEntry
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object ClusterShardAllocationMixin {
```

Comment:
This was refactoring of existing code to make use of the same `isAGoodTimeToRebalance`